### ISSUE TITLE
Entity.create with string relation typescript error

### ISSUE THREAD

**Issue by argaen** [typeorm/typeorm#10020]
### Issue description

Trying to use `Entity.create` with a relation being loaded as a string is showing typescript error

### Expected Behavior

The types defined are correct so I would expect to be able to create the entity with a one liner like:

```js
    account = await Account.create({
      guid: 'account_guid',
      name: 'name',
      type: 'ASSET',
      commodity: 'commodity_guid',
    }).save();
```

instead of having to do 

```js
    account = await Account.create({
      guid: 'account_guid',
      name: 'name',
      type: 'ASSET',
      commodity: commodityObject,
    }).save();
```

or 

```js
    account = Account.create({
      guid: 'account_guid',
      name: 'name',
      type: 'ASSET',
      commodity: 'commodity_guid',
    });

    account.commodity = 'commodity_guid'
    await account.save();
```

### Actual Behavior

Typescript is complaining with:

```bash
No overload matches this call.
  Overload 1 of 3, '(this: (new () => Account) & typeof BaseEntity, entityLikeArray: DeepPartial<Account>[]): Account[]', gave th
e following error.
    Argument of type '{ guid: string; name: string; commodity: string; }' is not assignable to parameter of type 'D
eepPartial<Account>[]'.
      Object literal may only specify known properties, and 'guid' does not exist in type 'DeepPartial<Account>[]'.
  Overload 2 of 3, '(this: (new () => Account) & typeof BaseEntity, entityLike: DeepPartial<Account>): Account', gave the followi
ng error.
    Type 'string' is not assignable to type 'DeepPartial<Commodity> | undefined'.
```

### Steps to reproduce

Create two basic entities like:

```js
export default class Account extends BaseEntity {
  @PrimaryColumn({
    type: 'varchar',
    length: 32,
  })
  @v.Length(1, 32)
    guid!: string;

  @Column({
    type: 'text',
    length: 2048,
  })
    name!: string;

  @ManyToOne('Commodity', { eager: true })
  @JoinColumn({ name: 'commodity_guid' })
  commodity!: Commodity;
}

export default class Commodity extends BaseEntity {
  @PrimaryColumn({
    type: 'varchar',
    length: 32,
  })
    guid!: string;

  @Column({
    type: 'text',
    length: 2048,
  })
    name!: string;
}
```

and then try to create an Account like this:

```js
    account = await Account.create({
      guid: 'account_guid',
      name: 'name',
      type: 'ASSET',
      commodity: 'commodity_guid',
    }).save();
```

I can solve the error by making the type of `commodity` in the Account entity `Commodity | string` but then, when accessing `accountInstance.commodity.name` it shows typescript error and I have to convert to `Commodity` manually every time. Like:

```
(accountInstance.commodity as Commodity).name
```

Is there a better way to do this or that's just it?

Thanks

### My Environment

typescript: 5.0.4
typeorm: 0.3.14
node: 16.17.0

### Relevant Database Driver(s)

- [ ] aurora-mysql
- [ ] aurora-postgres
- [ ] better-sqlite3
- [ ] cockroachdb
- [ ] cordova
- [ ] expo
- [ ] mongodb
- [ ] mysql
- [ ] nativescript
- [ ] oracle
- [ ] postgres
- [ ] react-native
- [ ] sap
- [ ] spanner
- [ ] sqlite
- [ ] sqlite-abstract
- [ ] sqljs
- [ ] sqlserver

### Are you willing to resolve this issue by submitting a Pull Request?

Yes, I have the time, but I don't know how to start. I would need guidance.

**Comment by argaen** [comment]
Hmmm a hacky solution (I believe) is doing this:

```
  @ManyToOne('Commodity', { eager: true })
  _commodity!: Commodity | string;

  get commodity(): Commodity {
    return this._commodity as Commodity;
  }
```

and then use `_commodity` when creating but `commodity` when accessing values. Typescript will let you know if you are accessing/setting the wrong thing but still... feels quite hacky

**Comment by piotrczech** [comment]
I encountered the same issue when working with ManyToOne relationships where only identifiers (id or uuid) are available at runtime, such as:

```ts
await repository.save({
  system: someSystemUuid,
  item: someItemId,
});
```

While this works fine at runtime, TypeScript raises errors like:

```
Type 'string' is not assignable to type 'System'.ts(2322) ItemExternalId.ts(19, 5): The expected type comes from property 'system' which is declared here on type 'Partial<ItemExternalId>'
Type 'number' is not assignable to type 'Item'.ts(2322) ItemExternalId.ts(14, 5): The expected type comes from property 'item' which is declared here on type 'Partial<ItemExternalId>'
```

The workaround suggested in this issue involves changing the type definitions of the Relation<T> or using additional type mappings. However, these solutions are not ideal because they require altering the core behavior or introducing unnecessary complexity into the type system.

Is there any progress on addressing this limitation, or could we consider reopening this issue for further discussion?

**Comment by tskorupka** [comment]
Hey, 

I got same case, solved by

from:
```
public async create(
    id: string,
    text: string,
    userId: string,
  ): Promise<void> {
    const newEntityInstance = this.otherRepository.create({
      id: randomUUID(),
      text,
      relationshipA: await this.repository.findOneBy({ id }),
      createdBy: userId,
    });

    await this.otherRepository.save(newEntityInstance);
  }
```

to
```
public async create(
    id: string,
    text: string,
    userId: string,
  ): Promise<void> {
    const newEntityInstance = this.otherRepository.create({
      id: randomUUID(),
      text,
      relationshipA: id as any as RelationshipA,
      createdBy: userId,
    });

    await this.otherRepository.save(newEntityInstance);
  }
```

**Comment by strace466** [comment]
> Hey,
> 
> TL;DR; - You can just pass a uuid as string or any other pk, cast it to any type, and then cast it to the one expected by your entity.
> 
> I got same case, solved by, e.g.
> 
> from:
> 
> ```
> public async create(
>     id: string,
>     text: string,
>     userId: string,
>   ): Promise<void> {
>     const newEntityInstance = this.otherRepository.create({
>       id: randomUUID(),
>       text,
>       relationshipA: await this.repository.findOneBy({ id }),
>       createdBy: userId,
>     });
> 
>     await this.otherRepository.save(newEntityInstance);
>   }
> ```
> 
> to
> 
> ```
> public async create(
>     id: string,
>     text: string,
>     userId: string,
>   ): Promise<void> {
>     const newEntityInstance = this.otherRepository.create({
>       id: randomUUID(),
>       text,
>       relationshipA: id as any as RelationshipA,
>       createdBy: userId,
>     });
> 
>     await this.otherRepository.save(newEntityInstance);
>   }
> ```





https://typescripter.space/questions/the-configuration-file-for-typescript-and-typeorm-specifically-the-ts-file-is-encountering-an-error

**Comment by tskorupka** [comment]
@strace466 I do not understand, why toy have posted some url which is unrelated?

**Comment by piotrczech** [comment]
This issue should not be closed, as the bug still occurs. The behavior is well documented, there are indications of a possible fix, and the problem seems significant since it has been reported by multiple users
